File: frontend/src/types/resources.ts
export interface Resource {
  arn: string;
  resource_type: string;
  name: string;
  region: string;
  tags: Record<string, string>;
  creation_date: string | null;
  iam_role: string | null;
  attributes: Record<string, unknown>;
  is_external: boolean;
  is_unresolved: boolean;
}

export interface Relationship {
  source_arn: string;
  target_arn: string;
  category: "network" | "iam" | "event" | "data";
  derived_from: string;
}

export interface RegionFailure {
  region: string;
  resource_type: string;
  error_message: string;
  timestamp: string;
}

export interface ScanResult {
  account_id: string;
  scan_timestamp: string;
  resources: Resource[];
  relationships: Relationship[];
  failures: RegionFailure[];
  scanned_regions: string[];
  total_scan_duration_ms: number;
}

File: backend/models/resources.py
"""Pydantic models for AWS resources and relationships."""

from typing import Any, Literal

from pydantic import BaseModel


class Resource(BaseModel):
    """Represents a single discovered AWS resource."""

    arn: str
    resource_type: str  # e.g., "ec2", "lambda", "s3"
    name: str
    region: str
    tags: dict[str, str] = {}
    creation_date: str | None = None  # ISO 8601
    iam_role: str | None = None
    attributes: dict[str, Any] = {}  # service-specific metadata
    is_external: bool = False
    is_unresolved: bool = False


class Relationship(BaseModel):
    """A detected connection between two AWS resources."""

    source_arn: str
    target_arn: str
    category: Literal["network", "iam", "event", "data"]
    derived_from: str  # configuration property name
